training.py

```diff
@@ -1,7 +1,67 @@
-# Somehow need to train agents using the genetic algorithm. This should probably be a
-# script that takes some sort of number of training iterations and what chromosomes to
-# use as parameters
+import pygad
+import numpy as np
+import logging
+import os
+from datetime import datetime
 
+# Function to set up logging
+def setup_logging():
+    log_folder = "genetic_algorithm_logs"
+    os.makedirs(log_folder, exist_ok=True)
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+    log_filename = os.path.join(log_folder, f'ga_log_{timestamp}.txt')
+    logging.basicConfig(filename=log_filename, level=logging.INFO, format='%(asctime)s - %(message)s')
+
+# Fitness function
+def fitness_function(ga_instance, solution, solution_idx):
+    fitness = np.sum(solution)  # Placeholder fitness function
+    return fitness
+
+# Callback function called after each generation
+def on_generation(ga_instance):
+    population = ga_instance.population
+    fitness_values = ga_instance.last_generation_fitness
+
+    logging.info(f'Generation {ga_instance.generations_completed}')
+    logging.info(f'Population:\n{population}')
+    logging.info(f'Fitness values: {fitness_values}\n')
+
+# Runs the genetic algorithm
+def run_genetic_algorithm(chromosome_length, population_size, num_generations, num_parents_mating):
+    population = np.random.randint(2, size=(population_size, chromosome_length))
+
+    logging.info(f'Initial population:\n{population}')
+
+    # Compute and log the initial fitness values
+    initial_fitness_values = [fitness_function(None, population[i], i) for i in range(population_size)]
+    logging.info(f'Initial fitness values: {initial_fitness_values}\n')
+
+    ga_instance = pygad.GA(
+        num_generations=num_generations,
+        num_parents_mating=num_parents_mating,
+        fitness_func=fitness_function,
+        sol_per_pop=population_size,
+        num_genes=chromosome_length,
+        initial_population=population,
+        gene_type=int,  # Ensures genes remain integers
+        mutation_percent_genes=10,  # Set mutation percentage
+        on_generation=on_generation  # Callback function for logging each generation
+    )
+
+    ga_instance.run()
+    ga_instance.plot_fitness()
+
+    solution, solution_fitness, solution_idx = ga_instance.best_solution()
+    logging.info(f'Best solution: {solution}')
+    logging.info(f'Best solution fitness: {solution_fitness}')
+
+# Main execution block
 if __name__ == "__main__":
-    # do script things
-    pass
+    setup_logging()
+
+    CHROMOSOME_LENGTH = 30 
+    POPULATION_SIZE = 30  
```

Comment:
Might be nice if some of the parameters we'd want to mess around with (generations, pop size) could be parsed from the command line. Forgive me if you already know this, but the argparse library is great for this and it lets you set defaults arguments for things you don't want to specify every time

Reply:
I agree, command line arguments are probably better for this. Will do!